### ISSUE TITLE
Efficiency of covariance calculation with high dimensionality

### ISSUE THREAD

**Issue by Giodiro** [getkeops/keops#57]
Hi,
First of all thank you for this amazing library.

I encountered a behaviour which was unexpected on my part and was wondering if there is anything I can do to fix it.
When calculating a kernel between `N x D` matrices with large `D` keops seems to slow down a lot.

Simple code sample to reproduce this is:
```python
import torch
from pykeops.torch import Genred
import timeit

a = torch.randn(10000, 700, requires_grad=False, dtype=torch.float64)
c = torch.randn(10000, 700, requires_grad=False, dtype=torch.float64)
v = torch.randn(10000, 2, requires_grad=False, dtype=torch.float64)

formula = '(X|Y) * v'
aliases = [
    'X = Vi(%d)' % (a.shape[1]),
    'Y = Vj(%d)' % (c.shape[1]),
    'v = Vi(%d)' % (v.shape[1]),
]
mmv = Genred(formula, aliases, reduction_op='Sum', axis=1, dtype='float64')
mmv(a, c, v)

timeit.repeat("mmv(a, c, v, backend='GPU_1D'); torch.cuda.synchronize()", globals=globals(), number=1, repeat=5)
timeit.repeat('(a @ c.T) @ v', globals=globals(), number=1, repeat=5)
```

The keops function takes ~6 seconds to run (on the GPU) while the naive pytorch takes ~0.4s (on a 24-core CPU). I find this interesting since if we reduce D to e.g. 7 KeOps is massively faster!

I'm sure there is something simple that I am clearly missing. Please let me know if this is the case.
Thanks,
Giacomo

**Comment by bcharlier** [comment]
Hi @Giodiro ,

thank for your messages. Two remarks:

1) there is a typo in your example (v should be a `j` variable). Namely:
```python
formula = '(X|Y) * v'
aliases = [
    'X = Vi(%d)' % (a.shape[1]),
    'Y = Vj(%d)' % (c.shape[1]),
    'v = Vj(%d)' % (v.shape[1]),
]
```

2) you don't need pykeops in your example: with the notations of your example,  doing `a @ (c.T @ v)` prevent you to store a `10000 x 10000` matrix. My  guess is that the torch parser [realized this](https://pytorch.org/docs/stable/torch.html#torch.chain_matmul) for you...

**Comment by Giodiro** [comment]
Hi @bcharlier , thanks for your answer and sorry for the typo in the variable v.

I understand that pytorch might be optimizing out the creation of the kernel but I get very similar results with:
```python
timeit.repeat("k = a@c.T; print(k.sum()); out = k @ v", globals=globals(), number=1, repeat=5)
```

I agree that pykeops should not improve things much in this case: it was just to exemplify the problem. For a Gaussian kernel (i.e. `'Exp(g * SqDist(X, Y)) * v'`) I get the same behaviour where a small `D` is extremely fast, but as `D` increases pykeops slows down a lot (compared to a torch implementation).

I understand this might be expected behaviour from the library; if so I would be interested in understanding why and if it can be optimized.

Thanks again!

**Comment by libkeops** [comment]
Hi @Giodiro ,

in fact, this remind me another story where the problems came from  the lack of `#pragma  unroll` in the code. By default, `nvcc` unrolls `for` loops (with a number `d` of loops known at compile time) up to `d=100`... So we aggressively ask to unroll everything (see e.g. [here](https://github.com/getkeops/keops/blob/master/keops/core/formulas/maths/Scal.h#L45))

I have performed a simple test (with a non linear kernel as you did). 
```python
import numpy as np
import torch
from pykeops.torch import Genred
import timeit

import matplotlib.pyplot as plt
D_list = np.arange(0,3000,100)
times_list = np.zeros(len(D_list))
times_t_list = np.zeros(len(D_list))
i= -1

for D in D_list:
    i += 1
    d2 = D*D
    a = torch.randn(10000, D, requires_grad=False, dtype=torch.float32).to('cuda')
    c = torch.randn(10000, D, requires_grad=False, dtype=torch.float32).to('cuda')
    v = torch.randn(10000, 2, requires_grad=False, dtype=torch.float32).to('cuda')

    formula = 'Exp((X|Y)/IntCst(' + str(d2) + ')) * v'
    aliases = [
        'X = Vi(%d)' % (a.shape[1]),
        'Y = Vj(%d)' % (c.shape[1]),
        'v = Vj(%d)' % (v.shape[1]),
    ]
    mmv = Genred(formula, aliases, reduction_op='Sum', axis=1, dtype='float32')
    t1 = mmv(a, c, v) # compilation

    times = timeit.repeat("mmv(a, c, v, backend='GPU_1D'); torch.cuda.synchronize()", globals=globals(), number=1, repeat=5)
    times_t = timeit.repeat('((a @ c.T /d2).exp()) @ v; torch.cuda.synchronize()', globals=globals(), number=1, repeat=5)

    times_list[i] = torch.tensor(times).mean().numpy()
    times_t_list[i] = torch.tensor(times_t).mean().numpy()

plt.plot(D_list, times_list, 'o-')
plt.plot(D_list, times_t_list, 'rx-')
plt.savefig("test_keops.png")

```

Here are the results for a size  `n=10000` (y-axis is the mean run time in seconds and x-axis is the dimension `d`). In blue : keops. In red pytorch. 

![pgf_fonts](https://user-images.githubusercontent.com/36916500/79323414-79ebfd00-7f0e-11ea-9c42-97f1a69c05cf.png)

Similar results for `n` up to 35000. Beyond this points there is an out-of-mem error. The situation in quite complex here as there is some constraints: the shared meme size (typically few kB per blocks), number of block, the parallelization schme to use. Thence, I put @jeanfeydy  and @joanglaunes  in the loop!

b.

**Comment by Giodiro** [comment]
Hi,
Thanks for the detailed reply. This interpretation makes a lot of sense, as I noticed that compilation time also increases with d.

**Comment by joanglaunes** [comment]
Hello,
We have now implemented a new scheme for dealing with large dimensions in KeOps. You can see the benchmark I just posted in issue #98 . Also here is the new output of the exact same code posted by Benjamin in April
(performed on a Titan Xp). As you can see runtimes are improved a lot (by a factor around 350 in some case !), yet we are still about 5 times slower than PyTorch unfortunately for this formula.

![test_keops](https://user-images.githubusercontent.com/34514333/95110596-8d01bc00-073e-11eb-996a-a015857e2763.png)
